# Basic Python

<div style="border:solid green 2px; padding: 20px"> <h1 style="color:green; margin-bottom:20px">Reviewer's comment v1</h1>

Hello Somil!

I'm happy to review your project today  🙌

You can find my comments under the heading **«Review»**. I will categorize my comments in green, blue or red boxes like this:

<div class="alert alert-success">
    <b>Success:</b> if everything is done successfully
</div>
<div class="alert alert-warning">
    <b>Remarks:</b> if I can give some recommendations or ways to improve the project
</div>
<div class="alert alert-danger">
    <b>Needs fixing:</b> if the block requires some corrections. Work can't be accepted with the red comments
</div>

Please don't remove my comments. If you have any questions, don't hesitate to respond to my comments in a different section.
<div class="alert alert-info"> <b>Student comments: </div>    


<div style="border:solid green 2px; padding: 20px">
<b>Reviewer's comment v1:</b>
    
<b>Overall Feedback</b> 
    
- Overall well done! I can see that a lot of effort has been made! Your project looks very good and you accomplished impressive results.
- However, there are some comments/areas left to fix that will help you to make your project even better (in red boxes with the title - `Reviewer's comment v1:`).
    
And of course, if you have any questions along the way, remember that you can always reach out to your tutor for any clarification.
    
I will wait for you to send me a new version of the project :)

    
</div>

<div style="border:solid green 2px; padding: 20px">
<b>Reviewer's comment v2:</b>
    
<b>Overall Feedback</b> 

Thank you for making all improvements in your project! It looks perfect now and there are no critical issues. 
    
Please keep up great work and don't hesitate to use this project as a reference in your future sprints. 
   
Good luck on the next project 🍀 
</div>

## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
It is always helpful for the reader to have additional information about project tasks. It gives an overview of what you are going to achieve in this project.


## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [30]:
# importing pandas
import pandas as pd

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [31]:
# reading the files and storing them to df
df= pd.read_csv('/datasets/movies_and_shows.csv')

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Well done! Required library has been imported.

Print the first 10 table rows:

In [32]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [33]:
# obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Great data overview and correct conclusions. 

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [34]:
# the list of column names in the df table
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [35]:
# renaming columns
df.columns= df.columns.str.strip().str.lower().str.replace(' ','_').str.replace('0','o')

Check the result. Print the names of the columns once more:

In [36]:
# checking result: the list of column names
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
It is great that you correctly applied `replace` function. You could also use `rename` function in that case.  

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [37]:
# calculating missing values
df.isnull().sum()/len(df)

name            0.000000
character       0.000000
role            0.000000
title           0.000012
type            0.000000
release_year    0.000000
genres          0.000000
imdb_score      0.053857
imdb_votes      0.055224
dtype: float64

<div class="alert alert-success">
<b>Reviewer's comment v1</b>
    
Great that you selected `isna()` method to find missing values! 

It is also sometimes helpful to check not only the total amount of missing values in each column but also look at the percentage of missing values. It helps to understand the overall impact. You can check percentage using, for example, this code:

`df.isnull().sum()/len(df)`

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [38]:
# dropping rows where columns with title, scores and votes have missing values
df=df.dropna()

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [39]:
# counting missing values
df.isnull().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Everything is correct here!

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [40]:
# counting duplicate rows
df.duplicated().sum()

6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [41]:
# Produce table with duplicates (with original rows included) and review last 5 rows
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
85574    False
85575    False
85576    False
85577     True
85578    False
Length: 80853, dtype: bool

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [42]:
# removing duplicate rows
df= df.drop_duplicates().reset_index(drop=True)

<div class="alert alert-warning">
<b>Reviewer's comment v1</b>
    
Note that it is also required additionally to re-create the indexes in your dataframe. To achieve that, you can use `reset_index()`.
    
Without that, the index column will not be ordinal anymore, as `drop_duplicates()` deleted some lines, hence the dataframe becomes less informative.
    
    
You can read about it here: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html
    
And parameter `drop=True` could be helpful to rewrite existent index instead of creating a second ordinal column.

Check for duplicate rows once more to make sure you have removed all of them:

In [43]:
# checking for duplicates
df.duplicated().sum()

0

<div class="alert alert-success">
<b>Reviewer's comment v1</b>
    
You correctly found and deleted all duplicates from the data.

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [44]:
# viewing unique type names
sorted(df['type'].unique())

['MOVIE',
 'SHOW',
 'movies',
 'shows',
 'the movie',
 'tv',
 'tv series',
 'tv show',
 'tv shows']

Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [25]:
# function for replacing implicit duplicates
def replace_wrong_show(df, wrong_shows_list,correct_show):
    df['type'] =df['type'].replace(wrong_shows_list,correct_show)
    return df

Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [50]:
wrong_shows_list = ['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv']
correct_show = 'show'
replace_wrong_show(df,'wrong_shows_list','correct_show')

name                character      role        title  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
73854       A??da Morales                  Maritza     ACTOR      Lokillo   
73855    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
73856  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
73857        In??s Prieto                    Fanny     ACTOR      Lokillo   
73858      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            type  release_year              genres  imdb_score  imdb_votes  
0          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
1          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
2          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
3          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
4          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
...          ...           ...                 ...         ...         ...  
73854  the movie          2021          ['comedy']         3.8        68.0  
73855  the movie          2021          ['comedy']         3.8        68.0  
73856  the movie          2021          ['comedy']         3.8        68.0  
73857  the movie          2021          ['comedy']         3.8        68.0  
73858  the movie          2021          ['comedy']         3.8        68.0  

[73859 rows x 9 columns]

In [58]:
# removing implicit duplicates
df.drop_duplicates()

name                character      role        title  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
73854       A??da Morales                  Maritza     ACTOR      Lokillo   
73855    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
73856  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
73857        In??s Prieto                    Fanny     ACTOR      Lokillo   
73858      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            type  release_year              genres  imdb_score  imdb_votes  
0          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
1          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
2          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
3          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
4          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
...          ...           ...                 ...         ...         ...  
73854  the movie          2021          ['comedy']         3.8        68.0  
73855  the movie          2021          ['comedy']         3.8        68.0  
73856  the movie          2021          ['comedy']         3.8        68.0  
73857  the movie          2021          ['comedy']         3.8        68.0  
73858  the movie          2021          ['comedy']         3.8        68.0  

[73859 rows x 9 columns]

<div class="alert alert-danger">
<b>Reviewer's comment v1</b>
    
Here you need to provide a list named `wrong_shows_list` while you have `wrong`.
Could you please adjust this small typo? 

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [59]:
# viewing unique genre names
df['type'].unique()

array(['MOVIE', 'the movie', 'tv show', 'shows', 'movies', 'tv shows',
       'tv series', 'tv', 'SHOW'], dtype=object)

<div class="alert alert-success">
<b>Reviewer's comment v2</b>
    
Your function works perfectly fine.
    

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [60]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name
new=df.loc[df['release_year']>=1999]
new

name            character      role     title       type  \
1661          Jeff Probst       Himself - Host     ACTOR  Survivor   tv shows   
1952     Benicio del Toro  Franky Four Fingers     ACTOR    Snatch      MOVIE   
1953        Dennis Farina           Cousin Avi     ACTOR    Snatch      MOVIE   
1954         Vinnie Jones    Bullet Tooth Tony     ACTOR    Snatch      MOVIE   
1955            Brad Pitt        Mickey O'Neil     ACTOR    Snatch      MOVIE   
...                   ...                  ...       ...       ...        ...   
73854       A??da Morales              Maritza     ACTOR   Lokillo  the movie   
73855    Adelaida Buscato           Mar??a Paz     ACTOR   Lokillo  the movie   
73856  Luz Stella Luengas         Karen Bayona     ACTOR   Lokillo  the movie   
73857        In??s Prieto                Fanny     ACTOR   Lokillo  the movie   
73858      Julian Gaviria              unknown  DIRECTOR   Lokillo  the movie   

       release_year               genres  imdb_score  imdb_votes  
1661           2000          ['reality']         7.4     24687.0  
1952           2000  ['crime', 'comedy']         8.3    841435.0  
1953           2000  ['crime', 'comedy']         8.3    841435.0  
1954           2000  ['crime', 'comedy']         8.3    841435.0  
1955           2000  ['crime', 'comedy']         8.3    841435.0  
...             ...                  ...         ...         ...  
73854          2021           ['comedy']         3.8        68.0  
73855          2021           ['comedy']         3.8        68.0  
73856          2021           ['comedy']         3.8        68.0  
73857          2021           ['comedy']         3.8        68.0  
73858          2021           ['comedy']         3.8        68.0  

[69881 rows x 9 columns]

In [61]:
# repeat conditional indexing so df has only shows (movies are removed as result)
new= new.loc[df['type'].str.upper()=='SHOW']
new

name character      role        title  type  \
3636     Tracee Ellis Ross   unknown     ACTOR  Girlfriends  SHOW   
3637         Golden Brooks   unknown     ACTOR  Girlfriends  SHOW   
3638          Persia White   unknown     ACTOR  Girlfriends  SHOW   
3639          Reggie Hayes   unknown     ACTOR  Girlfriends  SHOW   
3640      Jill Marie Jones   unknown     ACTOR  Girlfriends  SHOW   
...                    ...       ...       ...          ...   ...   
73614     Maneerat Kam-Uan        Ae     ACTOR    Let's Eat  SHOW   
73615    Rudklao Amratisha   unknown     ACTOR    Let's Eat  SHOW   
73616      Jaturong Mokjok   unknown     ACTOR    Let's Eat  SHOW   
73617     Pisamai Wilaisak   unknown     ACTOR    Let's Eat  SHOW   
73618  Sarawut Wichiensarn   unknown  DIRECTOR    Let's Eat  SHOW   

       release_year               genres  imdb_score  imdb_votes  
3636           2000  ['drama', 'comedy']         7.2      4528.0  
3637           2000  ['drama', 'comedy']         7.2      4528.0  
3638           2000  ['drama', 'comedy']         7.2      4528.0  
3639           2000  ['drama', 'comedy']         7.2      4528.0  
3640           2000  ['drama', 'comedy']         7.2      4528.0  
...             ...                  ...         ...         ...  
73614          2021  ['drama', 'comedy']         8.2         5.0  
73615          2021  ['drama', 'comedy']         8.2         5.0  
73616          2021  ['drama', 'comedy']         8.2         5.0  
73617          2021  ['drama', 'comedy']         8.2         5.0  
73618          2021  ['drama', 'comedy']         8.2         5.0  

[13330 rows x 9 columns]

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

<div class="alert alert-success">
<b>Reviewer's comment v2</b>

You've filterd the data required data correctly. 


Additionally there are other methods that could be helpful to filter your data, for example:`query()`.

    
`df = df.query('type == "SHOW"')`
    
You can read about them additionally here: 
https://towardsdatascience.com/10-examples-that-will-make-you-use-pandas-query-function-more-often-a8fb3e9361cb
https://towardsdatascience.com/how-to-use-loc-and-iloc-for-selecting-data-in-pandas-bd09cb4c3d79    

In [62]:
# rounding column with scores
new.loc[:,'imdb_score'] = new['imdb_score'].round()
#checking the outcome with tail()
new.tail()

/opt/conda/envs/python3/lib/python3.9/site-packages/pandas/core/indexing.py:1676: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


name character      role      title  type  release_year  \
73614     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  SHOW          2021   
73615    Rudklao Amratisha   unknown     ACTOR  Let's Eat  SHOW          2021   
73616      Jaturong Mokjok   unknown     ACTOR  Let's Eat  SHOW          2021   
73617     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  SHOW          2021   
73618  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  SHOW          2021   

                    genres  imdb_score  imdb_votes  
73614  ['drama', 'comedy']         8.0         5.0  
73615  ['drama', 'comedy']         8.0         5.0  
73616  ['drama', 'comedy']         8.0         5.0  
73617  ['drama', 'comedy']         8.0         5.0  
73618  ['drama', 'comedy']         8.0         5.0

It is now time to identify outliers based on the number of votes.

In [64]:
# Use groupby() for scores and count all unique values in each group, print the result
new.groupby('imdb_score')['name'].nunique()


imdb_score
2.0       24
3.0       27
4.0      174
5.0      581
6.0     2365
7.0     4313
8.0     4182
9.0      531
10.0       8
Name: name, dtype: int64

<div class="alert alert-warning">
<b>Reviewer's comment v2</b>

You can remove duplicate columns in the following way: 
    
```
grouped_shows = df.groupby('imdb score')['name'].count()
```

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [65]:
# filter dataframe using two conditions (scores to be in the range 4-9)
filter = new[(new['imdb_score'] >= 4) & (new['imdb_score'] <= 9)]
# group scores and corresponding average number of votes, reset index and print the result
filtered = filter.groupby('imdb_score')['imdb_votes'].mean().reset_index()
filtered

imdb_score     imdb_votes
0         4.0    5296.553073
1         5.0    3159.726655
2         6.0    3503.104966
3         7.0    8701.014325
4         8.0   30323.097856
5         9.0  127101.043956

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [66]:
# round column with averages
filtered['imdb_votes']=filtered['imdb_votes'].round()
# rename columns
filtered.columns=['score','average_vote']
# print dataframe in descending order
final= filtered.sort_values(by='average_vote',ascending=False)
final

score  average_vote
5    9.0      127101.0
4    8.0       30323.0
3    7.0        8701.0
0    4.0        5297.0
2    6.0        3503.0
1    5.0        3160.0

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

<div class="alert alert-success">
<b>Reviewer's comment v2</b>

Everything is correct here. 

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.

<div class="alert alert-success">
<b>Reviewer's comment v2:</b>
 
Overall you did a great research.
    
A small tip regarding overall conclusion in the project: 
    It represents the overall work progress that you achieved. On a real project, this is probably the only thing the business will read. Therefore, it is crucial to indicate in a structured way all conclusions that you made on each step in the project.

For example:

- Replaced missing values in the following data with the following method.
- Replaced data types in the following columns.
- etc.
- We observe that ... factors impact ... 
- My analysis show ...
- I can recommend the following next steps / activities ...

It is also important to provide explanations and interpretations that will be interesting for business based on your analysis.

</div>